components/ai_chat/core/browser/engine/engine_consumer_oai.cc

```diff
@@ -338,4 +326,35 @@ void EngineConsumerOAIRemote::GetFocusTabs(const std::vector<Tab>& tabs,
   std::move(callback).Run(base::unexpected(mojom::APIError::InternalError));
 }
 
+base::Value::List EngineConsumerOAIRemote::BuildPageContentMessages(
+    const PageContentses& page_contents,
+    int max_associated_content_length,
+    int video_message_id,
+    int page_message_id) {
+  base::Value::List messages;
+
+  for (const auto& page_content : page_contents) {
+    std::string truncated_page_content =
+        page_content.get().content.substr(0, max_associated_content_length);
+    SanitizeInput(truncated_page_content);
+    max_associated_content_length -= truncated_page_content.size();
```

Comment:
since now page content is included separately, we don't need to substr each one every time, it should only be done when content size is larger than remaining max_associated_content_length.